src/ninja.cc

```diff
@@ -505,6 +507,220 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+  if (!node)
+    return;
+  Edge* edge = node->in_edge();
+  if (!edge)
+    return;
+  if (!seen_.insert(node).second)
+    return;
+
+  for (vector<Node*>::iterator in = edge->inputs_.begin();
+       in != edge->inputs_.end(); ++in) {
+    ProcessNode(*in);
+  }
+
+  string deps_type = edge->GetBinding("deps");
+  if (!deps_type.empty()) {
+    DepsLog::Deps* deps = deps_log_->GetDeps(node);
+    if (deps)
+      ProcessNodeDeps(node, deps->nodes, deps->node_count);
+  } else {
+    ImplicitDepLoader dep_loader(state_, deps_log_, disk_interface_);
+    vector<Node*> depfile_deps;
+    string depfile = edge->GetUnescapedDepfile();
+    if (depfile.empty())
+      return;
+    string err;
+    dep_loader.LoadDepfileDeps(edge, depfile, &depfile_deps, &err);
+    if (!depfile_deps.empty())
+      ProcessNodeDeps(node, &depfile_deps[0], depfile_deps.size());
+  }
+}
+
+void MissingDependencyScanner::ProcessNodeDeps(Node* node, Node** dep_nodes,
+                                               int dep_nodes_count) {
+  Edge* edge = node->in_edge();
+  set<Edge*> deplog_edges;
+  vector<Node*> deplog_nodes_without_generator;
+  for (int i = 0; i < dep_nodes_count; ++i) {
+    Node* deplog_node = dep_nodes[i];
+    Edge* deplog_edge = deplog_node->in_edge();
+    if (deplog_edge) {
+      deplog_edges.insert(deplog_edge);
+    } else if (IsNodeInGenDir(deplog_node)) {
+      nodes_missing_gen_.insert(deplog_node);
+      deplog_nodes_without_generator.push_back(deplog_node);
+    }
+  }
+  vector<Edge*> missing_deps;
+  for (set<Edge*>::iterator de = deplog_edges.begin(); de != deplog_edges.end();
+       ++de) {
+    if (!state_->PathExistsBetween(*de, edge)) {
+      missing_deps.push_back(*de);
+    }
+  }
+
+  if (!deplog_nodes_without_generator.empty()) {
+    for (vector<Node*>::iterator ni = deplog_nodes_without_generator.begin();
+         ni != deplog_nodes_without_generator.end(); ++ni) {
+      printf("No generator: %s uses %s, but I don't know who generates that.\n",
+           node->path().c_str(), (*ni)->path().c_str());
+    }
+  }
+  if (!missing_deps.empty()) {
+    set<string> missing_deps_rule_names;
+    for (vector<Edge*>::iterator ne = missing_deps.begin();
+         ne != missing_deps.end(); ++ne) {
+      for (int i = 0; i < dep_nodes_count; ++i) {
+        if (dep_nodes[i]->in_edge() == *ne) {
+          generated_nodes_.insert(dep_nodes[i]);
+          printf("Missing dep: %s uses %s (generated by %s)\n",
+              node->path().c_str(), dep_nodes[i]->path().c_str(),
+              (*ne)->rule().name().c_str());
+          generator_rules_.insert(&(*ne)->rule());
+          missing_deps_rule_names.insert((*ne)->rule().name());
+        }
+      }
+    }
+    missing_dep_path_count_ += missing_deps_rule_names.size();
+    for (set<string>::iterator re = missing_deps_rule_names.begin();
+         re != missing_deps_rule_names.end(); ++re) {
+      //printf("Error: %s lacks a dep path to %s\n", node->path().c_str(),
```

Comment:
remove or uncomment?

Reply:
obsolete after code move to src/missing_deps.cc (with cleanups like this)